Python/compile.c

```diff
@@ -6175,20 +6175,44 @@ compiler_pattern_mapping(struct compiler *c, pattern_ty p, pattern_context *pc)
     }
     // Collect all of the keys into a tuple for MATCH_KEYS and
     // COPY_DICT_WITHOUT_KEYS. They can either be dotted names or literals:
+
+    // will hold only Constant_kind keys
+    PyObject *seen = PySet_New(NULL);
+    if (seen == NULL) {
+        return -1;
+    }
+
     for (Py_ssize_t i = 0; i < size; i++) {
         expr_ty key = asdl_seq_GET(keys, i);
         if (key == NULL) {
             const char *e = "can't use NULL keys in MatchMapping "
                             "(set 'rest' parameter instead)";
             SET_LOC(c, ((pattern_ty) asdl_seq_GET(patterns, i)));
+            Py_DECREF(seen);
             return compiler_error(c, e);
         }
+
+        if (key->kind == Constant_kind) {
+            if (PySet_Contains(seen, key->v.Constant.value)) {
+                const char *e =  "Duplicate literal keys in match pattern are "
+                                 "not allowed";
```

Comment:
I'd like to make this error message the same as the runtime error message. We can also provide a hint to the user by clarifying *which* key is duplicated:

```suggestion
                const char *e =  "mapping pattern checks duplicate key (%R)";
```

You can then format the message by adding `key->v.Constant.value` to the `compiler_error` call below.